charts/DataTab.tsx

```diff
@@ -20,92 +20,9 @@ export class DataTab extends React.Component<{
-            d => d.property !== "color"
-        )
-        const uniqueEntitiesAcrossDimensions =
-            chart.sortedUniqueEntitiesAcrossDimensions
-
-        // only get days if chart has a day-indexed variable, else get years across dimensions
-        const indexingYears = sortBy(
-            dayIndexedCSV
-                ? yearIsDayVar?.yearsUniq
-                : uniq(flatten(dimensions.map(d => d.yearsUniq)))
-        )
-
-        const rows: string[] = []
-
-        const titleRow = ["Entity", "Code", dayIndexedCSV ? "Date" : "Year"]
-
-        dimensions.forEach(dim => {
-            if (this.isFixedYearDimension(dim)) titleRow.push("Year")
-            titleRow.push(csvEscape(dim.fullNameWithUnit))
-        })
-        rows.push(titleRow.join(","))
-
-        uniqueEntitiesAcrossDimensions.forEach(entity => {
-            indexingYears.forEach(year => {
-                const row: (string | number)[] = [
-                    entity,
-                    chart.entityMetaByKey[entity].code ?? "",
-                    chart.formatYearFunction(year)
-                ]
-
-                let rowHasSomeValue = false
-                dimensions.forEach(dim => {
-                    let value = null
-                    if (this.isFixedYearDimension(dim)) {
-                        const latestYearValue = dim.yearAndValueOfLatestValueforEntity(
-                            entity
-                        )
-                        if (latestYearValue) {
-                            row.push(
-                                dim.formatYear(first(latestYearValue) as number)
-                            )
-                            value = last(latestYearValue)
-                        } else row.push("")
-                    } else {
-                        value = dim.valueByEntityAndYear.get(entity)?.get(year)
-                    }
-
-                    if (value != null) {
-                        row.push(value)
-                        rowHasSomeValue = true
-                    } else row.push("")
-                })
-
-                // Only add rows which actually have some data in them
-                if (rowHasSomeValue) rows.push(row.map(csvEscape).join(","))
-            })
-        })
-
-        return new Blob([rows.join("\n")], { type: "text/csv" })
-    }
-
-    @computed get csvDataUri(): string {
-        return window.URL.createObjectURL(this.csvBlob)
-    }
-
-    @computed get csvFilename(): string {
-        return this.props.chart.data.slug + ".csv"
-    }
-
-    // IE11 compatibility
-    @action.bound onDownload(ev: React.MouseEvent<HTMLAnchorElement>) {
-        if (window.navigator.msSaveBlob) {
-            window.navigator.msSaveBlob(this.csvBlob, this.csvFilename)
-            ev.preventDefault()
-        }
-    }
-
     render() {
-        const { bounds, csvFilename } = this
+        const { bounds } = this
+        const csvGenerator = new CSVGenerator({ chart: this.props.chart })
```

Comment:
`render()` will potentially be called multiple times, creating a new instance of `CSVGenerator` each time. I think it's better to use this pattern here:

```ts
@computed get csvGenerator(): CSVGenerator {
	return new CSVGenerator(this.props.chart)
}
```

This way, mobx will memoize the `CSVGenerator` until the `chart` prop changes (since it never does, it should always create only a single instance).

Reply:
And this is not relevant here, but, if you want to "force" there to be a single instance even if a property changes, you can use this pattern:

https://github.com/owid/owid-grapher/blob/89e6f4bd792ded8381b9b0da947b35240983b750/charts/LineChart.tsx#L104-L117

This took me a long while to understand – it circumvents the mobx tracking because [mobx tracks property access](https://mobx.js.org/best/react.html#mobx-tracks-property-access-not-values), and we are passing getters which are not invoked in the computed method.

Reply:
Ah, (2) is an interesting method. I have seen it in several places and wondered what it was.
Good point on (1). I'll modify now.